src/net_processing.cpp

```diff
@@ -1781,16 +1794,19 @@ std::optional<std::string> PeerManagerImpl::FetchBlock(NodeId peer_id, const CBl
     // Ignore pre-segwit peers
     if (!CanServeWitnesses(*peer)) return "Pre-SegWit peer";
 
+    const uint256& hash{block_index.GetBlockHash()};
+
     LOCK(cs_main);
 
+    if (IsBlockRequestedFromPeer(hash, peer_id)) return "Already requested from this peer";
+
     // Forget about all prior requests
-    RemoveBlockRequest(block_index.GetBlockHash(), std::nullopt);
+    RemoveBlockRequest(hash, std::nullopt);
 
     // Mark block as in-flight
-    if (!BlockRequested(peer_id, block_index)) return "Already requested from this peer";
+    Assume(BlockRequested(peer_id, block_index));
```

Comment:
In 2b67ea46:

Instead of adding the new function, could just delete the `RemoveBlockRequest()` line and directly call `BlockRequested()`. Failing with a `Already requested from this peer` if the call return false.

`BlockRequested()` only fails when the block is already in-flight for that peer (it does the same as your new `IsBlockRequestedFromPeer()` function).